docs/proposals/20231227-koordlet-resctrl-qos-enhance.md

```diff
@@ -0,0 +1,468 @@
+
+- Resctrl runtime hook will create/update QoS class level by using Rule to subscribe NodeSLO
+- Resctrl runtime hook subscribe RunPodSandBox event, and will handle pod level LLC and MBA schemata init
+- Resctrl runtime hook subscribe CreateContainer event, and set closid to ContainerContext, ContainerContext will update OCI spec based on the closid
+- Resctrl runtime hook will register a callback to Reconciler to consume ContainerTaskIds and then update new ContainerTaskIds to corresponding resctrl control group and monitor group
+
+Concurrently with the implementation of the Resctrl runtime hook plugin, we will deploy a dedicated Resctrl metric collector. This collector leverages NodeSLO and PodsInformer to gather real-time data on fixed class and pod-level LLC and MBA. This architecture, as depicted in the below diagram, ensures comprehensive resource consumption insights, which are crucial for informing the dynamic adjustments.
+![Alt text](../images/resctrl.png)
+
+
+### User Stories
+
+#### Story 1
+As a cluster administrator, I want to apply and adjust LLC/MBA QoS class configuration during runtime to get better resource usage
+#### Story 2
+As a user, I want to adjust my workload's LLC/MBA resource during runtime.
+#### Story 3
+As a cluster administrator, I want to monitor cluster LLC/MBA resource usage.
+#### Story 4
+As a cluster administrator, I find some workloads are noisy neighbor, I want to limit these noisy neighbors' LLC/MBA.
+
+### Requirements (Optional)
+Need Koordinator to upgrade to 1.5.0+
+
+#### Functional Requirements
+
+Resctrl runtime hook plugin should support all existing functionalities by current Resctrl QoS plugin
+
+##### FR1
+
+##### FR2
+
+#### Non-Functional Requirements
+
+Non-functional requirements are user expectations of the solution. Include
+considerations for performance, reliability and security.
+
+##### NFR1
+
+##### NFR2
+
+### API
+To support pod level LLC/MBA limit, we add a new annotation, the annotation key is `node.koordinator.sh/resctrl`, the annotation value is defined as below:
+```go
+type Resctrl struct {
+    L3 map[int64]int64
+    MB map[int64]int64
+}
+```
+
+### Implementation Details/Notes/Constraints
+
+To achieve fine-grained control and monitoring of LLC and MBA resources, we propose the implementation of a two-pronged approach:
+1. Resctrl Runtime Hook Plugin
+2. Resctrl Metrics Collector
+
+**Proposed Implementation Steps**
+1. Migration of Existing QoS class cevel LLC/MBA function: Relocate current QoS class level Resctrl functions to the runtime hook plugin for more immediate execution.
+2. Implementation of QoS class level LLC/MBA metrics collector: Introduce a dedicated monitor within the metrics collector to track resource usage at the QoS class level.
+3. Enabling Pod Level Resctrl Support: Extend the runtime hook plugin and metrics collector to encompass both control and monitoring capabilities at the pod level.
```

Comment:
it will be better if the pod level control can be implemented as the first since it is a new feature
pod level monitoring can be set as the last one.

Reply:
yes, already modified